File: VB/ImageStreamProvider.vb
Imports System.IO
Imports System.Data
Imports DevExpress.Office.Services

Namespace RichEditImageMailMerge

'#Region "#iuristreamprovider"
    Public Class ImageStreamProvider
        Implements IUriStreamProvider

        Private Shared ReadOnly prefix As String = "dbimg://"

        Private table As DataTable

        Private columnName As String

        Public Sub New(ByVal sourceTable As DataTable, ByVal imageColumn As String)
            table = sourceTable
            columnName = imageColumn
        End Sub

        Public Function GetStream(ByVal uri As String) As Stream Implements IUriStreamProvider.GetStream
            uri = uri.Trim()
            If Not uri.StartsWith(prefix) Then Return Nothing
            Dim strId As String = uri.Substring(prefix.Length).Trim()
            Dim id As Integer
            If Not Integer.TryParse(strId, id) Then Return Nothing
            Dim row As DataRow = table.Rows.Find(id)
            If row Is Nothing Then Return Nothing
            Dim bytes As Byte() = TryCast(row(columnName), Byte())
            If bytes Is Nothing Then Return Nothing

            Dim memoryStream As MemoryStream = New MemoryStream(bytes)
            Return memoryStream
        End Function
'#End Region  ' #iuristreamprovider
    End Class
End Namespace

File: CS/ImageStreamProvider.cs
using System.IO;
using System.Data;
using DevExpress.Office.Services;

namespace RichEditImageMailMerge {
#region #iuristreamprovider
    public class ImageStreamProvider : IUriStreamProvider {
        static readonly string prefix = "dbimg://";
        DataTable table;
        string columnName;

        public ImageStreamProvider(DataTable sourceTable, string imageColumn) {
            this.table = sourceTable;
            this.columnName = imageColumn;
        }


        public Stream GetStream(string uri) {
            uri = uri.Trim();
            if (!uri.StartsWith(prefix))
                return null;
            string strId = uri.Substring(prefix.Length).Trim();
            int id;
            if (!int.TryParse(strId, out id))
                return null;
            DataRow row = table.Rows.Find(id);
            if (row == null)
                return null;
            byte[] bytes = row[columnName] as byte[];
            if (bytes == null)
                return null;

            MemoryStream memoryStream = new MemoryStream(bytes);
            return memoryStream;
        }

#endregion #iuristreamprovider
    }
}